staging/src/k8s.io/apiserver/pkg/cel/library/image.go

```diff
@@ -0,0 +1,317 @@
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+const (
+	// DefaultTag is the tag name that will be used if no tag provided
+	DefaultTag = "latest"
+)
+
+// Image provides a CEL function library extension for parsing image references.
+//
+// image
+//
+// Converts a string to an image reference or returns an error if the image is not a valid image reference. Refer
+// to github.com/opencontainers/image-spec documentation for information on accepted image references.
+// An optional "normalize" argument can be passed to enable normalization. Normalization parses the image
+// into a normalized image. An example normalized image is "docker.io/library/ubuntu" for image "ubuntu".
+//
+//	image(<string>) <Image>
+//
+// Examples:
+//
+//	image('registry.k8s.io/kube-apiserver-arm64:latest') // Returns an Image
+//	image('nginx') // Returns an Image
+//	image('registry.k8s.io/kube-apiserver-arm64@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns an Image
+//	image('registry.k8s.io/kube-apiserver-arm64:latest@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns an Image
+//	image('registry.k8s.io//kube-apiserver-arm64') // error
+//	image(0) // error
+///	image('ubuntu') // Returns Image for "ubuntu".
+///	image('ubuntu', true) // Applies normalization and returns components of "docker.io/library/ubuntu".
+///	image('ubuntu', false) // Returns Image for "ubuntu".
+//	image('registry.k8s.io/kube-apiserver-arm64:latest', true) // Returns an Image
+//
+// isImage
+//
+// Returns true if a string is a valid image. isImage returns true if and
+// only if image reference parsing does not result in error.
+//
+//	isImage(<string>) <bool>
+//
+// Examples:
+//
+//	isImage('registry.k8s.io/kube-apiserver-arm64:latest') // Returns true
+//	isImage('nginx') // Returns true
+//	isImage('registry.k8s.io/kube-apiserver-arm64@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns true
+//	isImage('registry.k8s.io/kube-apiserver-arm64:latest@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns true
+//	isImage('registry.k8s.io//kube-apiserver-arm64') // Returns false
+//	isImage(0) // Returns false
+//
+// <Image>.containsDigest:
+//
+// Returns true if the Image has a digest.
+//
+// Examples:
+//
+//	image("image("registry.k8s.io/kube-apiserver-arm64:latest@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").containsDigest()").containsDigest() // Returns true
+//	image("registry.k8s.io/kube-apiserver-arm64@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").containsDigest()) // Returns true
+//	image("registry.k8s.io/kube-apiserver-arm64:latest").containsDigest() // Returns false
+//	image("registry.k8s.io/kube-apiserver-arm64").containsDigest() // Returns false
+//
+// Image components:
+//
+//   - registry: Returns the registry component of the image reference.
```

Comment:
Is registry port supported?  Examples of that would be good to see

Reply:
Runtimes support it, and IIRC this technically makes parsing ambiguous https://kind.sigs.k8s.io/docs/user/local-registry/